### PR TITLE
drivers: dma_sam0: support block chaining

### PR DESCRIPTION
**Pull request by TomFinet**
Previously, only transactions with a single block descriptor were supported. This commit adds and configures block descriptors that are linked together using memory from a statically allocated descriptor pool.

It also isolates the logic for configuring a block descriptor by placing it in its own function.

Note: this is a first draft, criticism is welcome!

### CODE REVIEW

**Comment by TomFinet** [comment]
Will continue work on this over the weekend.

**Comment by nandojve** [comment]
Hi @TomFinet ,

Let us know how is your progress on this.

**Comment by TomFinet** [comment]
Hi @nandojve, I got busy and haven't worked on fixing the failing checks yet. I can give it a look this weekend.

**Comment by nandojve** [comment]
Hi @TomFinet ,

Would you mind to rebase on top of main and force a push to see if get CI green : ) ?

**Comment by nandojve** [comment]
Hi @TomFinet ,

In general LGTM! I was wondering to know how have you been testing?

**Comment by TomFinet** [comment]
I was thinking of writing a `chan_chain_transfer` test in `tests/drivers/dma` to test a dma transfer with chaining. I can make a new commit of this tomorrow.

File: tests/drivers/dma/chan_chain_transfer/boards/atsamc21n_xpro.overlay [hunk 1]
```diff
@@ -0,0 +1,7 @@
+/*
+ * Copyright (c) 2022 Kumar Gala <galak@kernel.org>
```
Comment (nandojve) [inline tests/drivers/dma/chan_chain_transfer/boards/atsamc21n_xpro.overlay:2]: I think you can keep only your name in those.
Please, add samr21 in the list.

File: tests/drivers/dma/chan_chain_transfer/src/main.c [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright (c) 2020 NXP
```
Comment (nandojve) [inline tests/drivers/dma/chan_chain_transfer/src/main.c:2]: ditto

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @TomFinet ,

I tested on samr21 and got below:

```
...

E: Channel number must be greater than any existing active channel numbers that use linked descriptors.
ERROR: transfer

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/dma/chan_chain_transfer/src/test_dma.c:152: dma_chain_m2m_test_test_dma0_chain_m2me

...

SUITE FAIL -  50.00% [dma_chain_m2m]: pass = 2, fail = 2, skip = 0, total = 4 duration = 4.116 seconds
 - PASS - [dma_chain_m2m.test_test_dma0_chain_m2m_chan0_burst16] duration = 2.019 seconds
 - FAIL - [dma_chain_m2m.test_test_dma0_chain_m2m_chan0_burst8] duration = 0.039 seconds
 - PASS - [dma_chain_m2m.test_test_dma0_chain_m2m_chan1_burst16] duration = 2.019 seconds
 - FAIL - [dma_chain_m2m.test_test_dma0_chain_m2m_chan1_burst8] duration = 0.039 seconds
```

Just a reminding that samr21 is in fact samd21 + rf233 radio.

**Pull request closed without merging**